Virtual Router Redundancy Protocol Adaptation HLD

micas-net opened the pull request:
Hld for VRRP adaptation to SONiC, as mentioned in issue: https://github.com/sonic-net/SONiC/issues/1366:


| Repo  | PR title | State |
| ----------------- | ----------------- | ----------------- |
| sonic-buildimage | https://github.com/sonic-net/sonic-buildimage/pull/18617, https://github.com/sonic-net/sonic-buildimage/pull/20383 |
| sonic-swss | https://github.com/sonic-net/sonic-swss/pull/3106, https://github.com/sonic-net/sonic-swss/pull/3313, https://github.com/sonic-net/sonic-swss/pull/3315 |
| sonic-swss-common | https://github.com/sonic-net/sonic-swss-common/pull/813, https://github.com/sonic-net/sonic-swss-common/pull/927 |
| sonic-utilities | https://github.com/sonic-net/sonic-utilities/pull/2949 |

micas-net committed:
Create VRRP_Adaptation_HLD.md

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
update VRRP_Adaptation_HLD.md

micas-net committed:
update VRRP hld

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
update VRRP HLD

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
add yang modle

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
add yang model catalog

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
add load sharing unit test case

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
optimize vmac description

Signed-off-by: philo <philo@micasnetworks.com>

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,710 @@
+                    default 3；
+                }
+
+                leaf pre_empt {
```
> I suggest to change the name of this variable to `preempt_mode`.

thanks! will change to preempt_mode.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,710 @@
+                    type uint16 {
+                        range 1..254;
+                    }
+                    default 100;
```
> Default value means that if two routers will be deployed in the same LAN without specifying a priority option, both will be created with priority 100; this is absolutely not recommended and my opinion is that the priority should be a required argument and not have a default value.
> 
> Any other opinion about it?

thanks! if with same priority, the smaller ip route will be selected as master, but I think a mandatory type for this filed in yang is also fine.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,710 @@
+
+                leaf priority {
+                    type uint16 {
+                        range 1..254;
```
> I think that if SONiC does not support the `IP Address Owner`, also called pure Virtual IP system implementation, it should be documented in a dedicated section - such as: _**intentional limitations**_ or _**known limitations**_ - and the the section `Virtual Router Owner` of this HDL modified.

Thanks for reminder, will add such a session.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,710 @@
+- Linux kernel version 5.1+
+
+#### Scalability Requirements
+
```
> Should the maximum number of VIPs per instance also be a scalability requirement?

Yes, will put them together.

philo-micas commented:
> Hi all, adding the VRRP protocol implementation in SONiC would be a great improvement for me; really good job.
> 
> While reading the HLD, I left some comments (which for GitHub looks like a review); I hope those are not off-topics.

Thanks! The comments are helpful, wish for more reviewing comments.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,710 @@
+                }
+
+                leaf vrid {
+                    type uint16 {
```
> Could in this case `uint8` be more suitable?

yes.

micas-net committed:
update VRRP_Adaptation_HLD.md

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
Modify Segments

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
Update VRRP_Adaptation_HLD.md

prvattem commented:
@philo-micas There was a review request covering VRRP functionality earlier.
https://github.com/sonic-net/SONiC/pull/475
Please see if it can be combined with your proposal.

micas-net committed:
Update VRRP_Adaptation_HLD.md

philo-micas commented:
> @philo-micas There was a review request covering VRRP functionality earlier. #475 Please see if it can be combined with your proposal.

From the aspect of RFC, two proposals have already combined, but from up layer service, one is based on keepactived, the other is based on FRR, they can't be combined. such as version 3 supported, vtysh command and so on.

zhangyanzhao commented:
SONiC community review recording https://zoom.us/rec/share/_PaY38c6vF8N6iAWALtMg1Rt0d9Ou8Guf4_ZRNayRYobIPq0jryZv2AsRE-OlLx5.3O_GA7L8vNPjWhkQ

lguohan commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+  1. Virtual Router Owner
+  2. Accept Mode configurable function
+
+### Testing Requirements/Design  
```
Please also submit sonic-mgmt test plan for this feature.

lguohan commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+    - For VRRP instances state transition to Master, set the Macvlan device into protodown off
+    - For VRRP instances state transition to non-Master, set the Macvlan interface into protodown on
+
+vrrpmgrd:
```
vrrpmgrd, vrrpsyncd, vrrporch should be moved into swss container.

philo-micas replied:
OK.

lguohan commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+  - Listens to VRRP create, delete and parameter change in CONFIG DB
+  - Update changes by using FRR CLI commands
+
+vrrpd:
```
we should check why vrrpd cannot be moved into a separate container, and represent a standard feature.

philo-micas replied:
I think it's better to run all FRR components in one container, or other there may exist too many containers after the other FRR components are enabled in the future.

lguohan commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+
+#### CoPP Configurations
+
+CoPP will be extended as follows for trapping VRRPs:
```
the copp rules should be tied to the vrrp feature, it should not be enabled by default. the vrrp copp rules should only be enabled when vrrp feature is enabled.

philo-micas replied:
OK

lguohan commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+ * @default false
+ */
+
+SAI_ROUTER_INTERFACE_ATTR_IS_VIRTUAL,
```
please describe in hld if this attributed is not supported by a particular vendor sai implementation.

philo-micas replied:
ok

prvattem commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+"adv_interval"
+"1000"
+"state"
+""
```
Can the state be empty?

philo-micas replied:
No, I will correct it. Default is up. thanks!

venkatmahalingam commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+
+### Scope  
+
+This document describes the high level design of frr-vrrpd adaptation to SONiC.
```
Can you please capture alternative design with keepalived for VRRP? why is it important to choose FRR based VRRP rather keepalived based VRRP?

philo-micas replied:
yea, Keepalived and FRR are both routing software, and since FRR has already been choosed as the router protocol by SONiC long time ago, support muiltiple protocols except vrrp, such as BGP, OSPF, RIP, IS-IS, PIM, LDP, BFD,etc, 
	so I think no matter from the architecture, extension and aspect of opensource, and iteration speed, FRR will be more powerfull, and what more, above components have already be compiled in SONiC, just need to adapt and enable them.

venkatmahalingam commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+
+VRRP control packets have IP protocol type as 112 (reserved for VRRP), for IPv4 network, are sent to VRRP multicast address 224.0.0.18, source MAC in VRRP control packets is virtual MAC address and source IP is interface IP (not the virtual IPv4 address). For IPv6 networks, the source address in the VRRPv6 header is the IPv6 link local address of the sending interface, and the destination address ...
+
+##### Uplink Interface Tracking
```
I believe, this uplink interface tracking can be a standalone feature instead of combining as part of VRRP.

philo-micas replied:
yes, you are right, will consider to exact it out for a new feature "monitor link" in phase 2.

venkatmahalingam commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+    - Add/del Linux Macvlan device to kernel
+    - Config virtual MAC to Macvlan device;
+    - Add/del VIP to Macvlan device;
+    - Upadate VRRP instance configuration to the APPL DB.
```
Fix the typo.

philo-micas replied:
OK， thanks.

venkatmahalingam commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+  - Subscribes to CONFIG_DB tables, parsing configurations and passes to vrrpd by using FRR CLI commands. 
+
+vrrpd: 
+  - Responsible for all VRRP protocol related calculations. VRRP packets are sent and received in vrrpd and states are updated accordingly.
```
Per FRR documentation,
_FRR’s VRRP implementation uses Linux macvlan devices to to implement the shared virtual MAC feature of the protocol. Currently, it does not create those system interfaces - they must be configured outside of FRR before VRRP can be enabled on them._

Do you know whether FRR has any plan to create MACVLAN interface for VRRP in the near future?

Do we have any sequence issue that VRRP can be enabled in FRR only after the MACVLAN interface creation in the kernel?

philo-micas replied:
Have not found such plan, from the vrrpd user guider, need the user to do the creation of Macvlan device for different OS, since it depends on the kernel tool such as iptable2, some OS may not support that.
And  we have not faced such an issue about the sequence currently, and we will pay more attention about it later.

venkatmahalingam commented on doc/vrrp/VRRP_Adaptation_HLD.md:
````diff
@@ -0,0 +1,712 @@
+```
+admin@sonic:~$ redis-cli -n 0 HGETALL "VRRP_TABLE:Ethernet36:6:IPv4"
+1) "vmac"
+2) "00:00:5e:00:01:06"
````
When you have both vr-id and address family, why cant we derive the vmac in the orchagent internally?

philo-micas replied:
good suggestion!!

venkatmahalingam commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+
+### SAI API 
+
+VRRP virtual router interface can be created by setting the attribute “SAI_ROUTER_INTERFACE_ATTR_IS_VIRTUAL” with create_router_interface API.
```
How's VIP is being programmed into the HW?

venkatmahalingam commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+
+#### CLI/YANG Model Enhancements
+
+##### SNOiC VRRP YANG MODEL
```
Fix the typo.

philo-micas replied:
OK,thanks!

venkatmahalingam commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+                leaf-list vip {
+                    type union {
+                        type inet:ipv4-prefix;
+                        type inet:ipv6-prefix;
```
Why are we using prefix instead of inet:ip-address?

philo-micas replied:
There will some issue with prefix, because prefix will carry no mask, the 32-bit mask default routing causes ping failure issue.

venkatmahalingam commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+                    }
+                }
+
+                leaf-list track_interface {
```
Can we have the uplink interface tracking feature as a standalone one instead of combining this with VRRP feature?

philo-micas replied:
yes,  will consider to exact it out for a new feature "monitor link" in phase 2.

venkatmahalingam commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+
+                leaf version {
+                    type enumeration {
+                        enum 2;
```
We should have the validation to throw an error if IPv6 VIP is configured with version 2.

philo-micas replied:
OK

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+
+                key "interface_name vrid";
+
+                leaf interface_name {
```
ok.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+vip             = ip_address                    ; Virtual IP address. This is a list of IP addresses
+priority        = 3DIGIT                        ; Priority of VRRP instance
+adv_interval    = 5DIGIT                        ; Advertisement interval for VRRP. Default = 1000ms
+state           = vrrp_state                    ; String denoting the state of VRRP instance
```
will change state->admin_status, witch is used to control up/down status of instance, default is on.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+key         = VRRP_TABLE:interface_name:vrid:type
+              interface_name        ; interface name as a string. Vlan, Ethernet or PortChannel
+              vrid                  ; vrid is an integer
+              type                  ; IP address type string. IPv4 or IPv6
```
Currently, the type is only used for key to get entry, and will remove vmac field later in VRRP_TABLE, at that time type will be used for calculating the vmac in oa.

prvattem commented:
> Hi all, sorry to spam again, but I need to do it because it affects some of my previous comments (not reviews); I am only just aware that VRRP has a [RFC YANG data model](https://www.rfc-editor.org/rfc/rfc8347).
> 
> I found these sections that may cover other questions as well:
> 
> ```
> [...]
> 
>   /*  identity and its derivatives. */
>   identity vrrp-state-type {
>     description
>       "Indicates the state of a virtual router.";
>   }
>   identity initialize {
>     base vrrp-state-type;
>     description
>       "Indicates that the virtual router is waiting
>        for a startup event.";
>   }
>   identity backup {
>     base vrrp-state-type;
>     description
>       "Indicates that the virtual router is monitoring the
>        availability of the master router.";
>   }
>   identity master {
>     base vrrp-state-type;
>     description
>       "Indicates that the virtual router is forwarding
>        packets for IP addresses that are associated with
>        this virtual router.";
>   }
> 
> [...]
> 
>     leaf priority {
>       type uint8 {
>         range "1..254";
>       }
>       default 100;
>       description
>         "Configures the VRRP election priority for the backup
>          virtual router.";
>     }
> 
>     leaf is-owner {
>       type boolean;
>       config false;
>       description
>         "Set to 'true' if this virtual router is the owner.";
>     }
> 
>        leaf state {
>          type identityref {
>            base vrrp:vrrp-state-type;
>          }
>          config false;
>          description
>            "Operational state.";
>        }
> 
> [...]
> ```
> 
> I think that In the case of `preempt` and `priority`, @philo-micas's suggestions were certainly correct.
> 
> I also think we should try to follow the RFC 8347 as much as possible.
> 
> GitHub source: https://github.com/YangModels/yang/blob/main/standard/ietf/RFC/ietf-vrrp%402018-03-13.yang

A similar Yang model is also available in OpenConfig Yang, which can be taken as reference too:
https://github.com/openconfig/public/blob/5ef541dd3d468239f07d69fb05a4dbc9e3f7476b/release/models/interfaces/openconfig-if-ip.yang#L896

philo-micas commented:
> https://github.com/openconfig/public/blob/5ef541dd3d468239f07d69fb05a4dbc9e3f7476b/release/models/interfaces/openconfig-if-ip.yang#L896

OK

philo-micas commented:
> Hi all, sorry to spam again, but I need to do it because it affects some of my previous comments (not reviews); I am only just aware that VRRP has a [RFC YANG data model](https://www.rfc-editor.org/rfc/rfc8347).
> 
> I found these sections that may cover other questions as well:
> 
> ```
> [...]
> 
>   /*  identity and its derivatives. */
>   identity vrrp-state-type {
>     description
>       "Indicates the state of a virtual router.";
>   }
>   identity initialize {
>     base vrrp-state-type;
>     description
>       "Indicates that the virtual router is waiting
>        for a startup event.";
>   }
>   identity backup {
>     base vrrp-state-type;
>     description
>       "Indicates that the virtual router is monitoring the
>        availability of the master router.";
>   }
>   identity master {
>     base vrrp-state-type;
>     description
>       "Indicates that the virtual router is forwarding
>        packets for IP addresses that are associated with
>        this virtual router.";
>   }
> 
> [...]
> 
>     leaf priority {
>       type uint8 {
>         range "1..254";
>       }
>       default 100;
>       description
>         "Configures the VRRP election priority for the backup
>          virtual router.";
>     }
> 
>     leaf is-owner {
>       type boolean;
>       config false;
>       description
>         "Set to 'true' if this virtual router is the owner.";
>     }
> 
>        leaf state {
>          type identityref {
>            base vrrp:vrrp-state-type;
>          }
>          config false;
>          description
>            "Operational state.";
>        }
> 
> [...]
> ```
> 
> I think that In the case of `preempt` and `priority`, @philo-micas's suggestions were certainly correct.
> 
> I also think we should try to follow the RFC 8347 as much as possible.
> 
> GitHub source: https://github.com/YangModels/yang/blob/main/standard/ietf/RFC/ietf-vrrp%402018-03-13.yang

OK

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+
+Following requirements are addressed by the design presented in this document:
+  1. Support VRRPv2(IPv4) and VRRPv3(IPv4 and IPv6)
+  2. Support VRRP on Ethernet, VLAN and PortChannel interfaces
```
No, currently not supported.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+  3. Support interface configuration of multiple VRRP instances.
+  4. Support configurable priority for VRRP instance
+  5. Support configurable preempt mode for VRRP instance
+  6. Support uplink interface tracking feature
```
Currently not supported,

prsunny commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+##### SWSS container
+
+vrrporch: 
+  - Subscribes to APPL_DB tables, responsible for updating the ASIC DB. Creating/deleting Virtual RIF object via SAI API.
```
I dont see a need for separate vrrporch and needs to only handle few extra attributes for INTERFACE creation. Please consider to leverage intforch

philo-micas replied:
OK

vvbrcm commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+    - Update the kernel Macvlan device status to the APPL DB.
+    - Match VRRP instance tracking interface and recalculate priority.
+    - Update priority change to vrrpd by using FRR CLI commands.
+
```
Note that the track interface support should be added in VRRP in FRR. The tweaking of priority for interface tracking does not seem right. The interface status from kernel needs to be conveyed to VRRP and VRRP should adjust the priority. vrrpsyncd and vrrpcfg/frrcfg should not be involved for interface status change.

vvbrcm commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+2) "00:00:5e:00:01:06"
+3) "state"
+4) "up"
+5) "priority "
```
Please explain the need for priority. If not needed, please remove it.

philo-micas replied:
Good suggestion,thinks, we will remove it.

vvbrcm commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+admin@sonic:~$ redis-cli -n 0 HGETALL "VRRP_TABLE:Ethernet36:6:IPv4"
+1) "vmac"
+2) "00:00:5e:00:01:06"
+3) "state"
```
The set of VIPs that needs to be programed in hardware my IP table is missing.

philo-micas replied:
Good question, we are considering the scenario where the VIP and parent interface are in the same network segment. Therefore, whether the VIP is programmed to hardware does not seem to affect the VRRP feature.

vvbrcm commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+  4. Support configurable priority for VRRP instance
+  5. Support configurable preempt mode for VRRP instance
+  6. Support uplink interface tracking feature
+
```
Is sub-interface, VRF and owner case supported?

philo-micas replied:
Currently, the sub interface is waiting for validation; and VRF is supported;
due to causing lots of different issues, FRR has temporarily disabled the owner function.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,710 @@
+ Effective Priority (v4)              100                 
+ Effective Priority (v6)              100                 
+ Preempt Mode                         Yes                 
+ Accept Mode                          Yes                 
```
Thanks for reminder, will add such a session.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+##### BGP container
+
+vrrpcfgd: 
+  - Subscribes to CONFIG_DB tables, parsing configurations and passes to vrrpd by using FRR CLI commands. 
```
We need vrrpcfgd to obtain the relevant configuration of VRRP from config db.  CLI->CONFIG DB->vrrpcfgd->vrrpd.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,710 @@
+                    default 1000;
+                }
+
+                leaf state {
```
the “state” means admin_status, witch is used to control up/down status of instance, default is up. and will be modified state->admin_status.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+                                B is weight value, integer ranges from 10 to 50.
+                                for example, track_interface could be: Ethernet1|30,Ethernet2|40";
+
+                    type string;
```
Good suggestion! You're right, we will refer to OpenConfig Yang and modify it as follows:
```
leaf-list track_interface {
    type leafref {
        path /port:sonic-port/port:PORT/port:PORT_LIST/port:ifname;
    }
    max-elements 8;
    /* ordered-by user */
}

leaf weight {
    type uint8 {
        range 10..50;
    }
}
```
meanwhile, the config db will also make modifications.

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,712 @@
+
+Diagram 1. VRRP components.
+
+Significantly, This feature extends functionality implemented in [SONiC FRR-BGP Extended Unified Configuration Management Framework](https://github.com/sonic-net/SONiC/blob/master/doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md), Redis DB events will trigger frrcfgd when the field frr_mgmt_framework_config set to "true" in the DEVICE_METADATA table, otherwise will trigger vrrpcfgd, and th ...
```
We also implement support for VRRP in frrcfgd, frr_mgmt_framework_config is used to control frrcfg or vrrpcfgd, if set "true" in the DEVICE_METADATA table ,used frrcfgd; otherwise , used vrrpcfgd.

micas-net committed:
Updated the document based on some review comments from community members

CLI, yang, config db, appl db, requirement,etc

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
del VRRP_Vrrporch_Uml.png

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
update VRRP_Adaptation_HLD.md

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
update VRRP Architecture and copp Configurations

Signed-off-by: philo <philo@micasnetworks.com>

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,710 @@
+
+##### Pre-emption
+
+Toggle preempt mode. When enabled, preemption allows Backup routers with higher priority to take over Master status from the existing Master. Enabled by default.
```
Good suggestion! we will add，thanks!

philo-micas commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,710 @@
+                }
+
+                leaf priority {
+                    type uint16 {
```
Great suggestion, we will revise it. Thanks.

micas-net committed:
update APPL DB tables

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
Modifying the HLD based on the review comments

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
Modifying the HLD based on the review comments.

Signed-off-by: philo <philo@micasnetworks.com>

zhangyanzhao commented:
Thanks reviewers for the comments. Please help to approve this PR after your review. Thanks.

vvbrcm commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,739 @@
+#### CLI/YANG Model Enhancements
+
+##### SONiC VRRP YANG MODEL
+
```
Please use below file for VRRP sonic yang

https://github.com/dks19/SONiC/blob/master/doc/sonic-vrrp.yang

philo-micas replied:
OK ,thanks

vvbrcm commented on doc/vrrp/VRRP_Adaptation_HLD.md:
```diff
@@ -0,0 +1,739 @@
+  - Listens to Macvlan interface programming in kernel. 
+  - Update the kernel Macvlan device's state to the INTF_table entry of APPL DB.
+
+intforch:
```
Use vrrpsyncd and vrrporchd sections of VRRP_HLD

adyeung replied:
@philo-micas BRCM has met with Micas and agreed on a converged design on VRRP, please reference the slide deck and update the HLD based our last meeting. I am hoping to get another round of review from the community after your HLD update

philo-micas replied:
OK,thanks

vvbrcm commented on doc/vrrp/VRRP_Adaptation_HLD.md:
````diff
@@ -0,0 +1,739 @@
+```
+
+#### APPL_DB Changes
+
````
Please use sec 4.2.2. from VPPD_HLD (keepalived based).
The key should be base interface name (not vrrp), vip and type.
The value should have vmac.

philo-micas replied:
done,thanks

vvbrcm commented on doc/vrrp/VRRP_Adaptation_HLD.md:
````diff
@@ -0,0 +1,739 @@
+preempt               = "enabled"/"disabled"          ; VRRP pre-emption is enabled? Default is enabled
+track_interface       = interface_name                ; This is a list of configured tracking interfaces
+priority_decrement    = 2DIGIT                        ; Specifies how much to decrement the priority of the VRRP instance if the tracking interface goes down, Default is 20
+```
````
Please take the CONFIG_DB, APP_DB section from VRRP_HLD (keepalived based) sec 4.2.1

- Elaborate on track interface format
-

philo-micas replied:
done

micas-net committed:
Update VRRP HLD

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
Update VRRP_Architecture_Diagram.png

Signed-off-by: philo <philo@micasnetworks.com>

zhangyanzhao commented:
@philo-micas can you please help to check the code PRs? Will Micas target 202405 release for this feature or move to future release? Thanks.

micas-net committed:
Update VRRP YANG file from https://github.com/sonic-net/SONiC/blob/b13dd68c156e770fed05d305d4ad99accefdf142/doc/sonic-vrrp.yang

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
VRRP HLD revision contributed by Vijay-Broadcom

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
VRRP HLD fix format error in the Revision part.

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
VRRP HLD:
(1) Updated the HLD to reflect VRRP, VRRP6, VRRP_TRACK and VRRP6_TRACK tables.
(2) Removed the vrrpmgrd from the APP_DB VRRP_TABLE's Producers
(3) Updated the Architecture Diagram according to (2).

Signed-off-by: philo <philo@micasnetworks.com>

philo-micas commented:
> @philo-micas can you please help to check the code PRs? Will Micas target 202405 release for this feature or move to future release? Thanks.

 Code change will be ready soon, and will have a revivew at 3/28 Routing WG meeting.   Keep target 202405. Thanks.

micas-net committed:
Run vrrpmgrd in BGP container

Signed-off-by: philo <philo@micasnetworks.com>

micas-net committed:
change VIP type

Signed-off-by: philo <philo@micasnetworks.com>

philo-micas commented:
> Hi all, many thanks for that and for your work.
> 
> I just have one major concern regarding the actual and the future architecture: why is [`vrrpmgr`](https://github.com/philo-micas/sonic-swss/blob/d58a143c67099c4d73522b01b7bc7f220797e518/cfgmgr/vrrpmgr.cpp#L134-L154) adding and deleting VIPs into VMAC interfaces? This is not done by `FRR/vrrpd`? I don't know if the [code](https://github.com/philo-micas/sonic-swss/blob/d58a143c67099c4d73522b01b7bc7f220797e518/cfgmgr/vrrpmgr.cpp#L134-L154) is valid or ever triggered, but I don't see any reason to manage that part from the `vrrpmgr` and to put that code there.
> 
> Apart of that, and as also other users suggested, I also believe that warm reboot can be covered by this HLD.
> 
> Many thanks and regards.

When configuring VRRP instances, a mask needs to be carried to the database (the VIP attribute of config db needs to be carried with a mask), because when creating macvlan, configuring Vip requires a mask, otherwise it will default to a 32-bit mask, causing the ping Vip action to fail. And the configuration of macvlan VIP demonstrated in FRR's guidance document also carries a mask https://docs.frrouting.org/en/latest/vrrp.html .

micas-net committed:
renaming vrrpmgr into macvlanmgr

Signed-off-by: philo <philo@micasnetworks.com>

philo-micas commented:
> > > Hi all, many thanks for that and for your work.
> > > I just have one major concern regarding the actual and the future architecture: why is [`vrrpmgr`](https://github.com/philo-micas/sonic-swss/blob/d58a143c67099c4d73522b01b7bc7f220797e518/cfgmgr/vrrpmgr.cpp#L134-L154) adding and deleting VIPs into VMAC interfaces? This is not done by `FRR/vrrpd`? I don't know if the [code](https://github.com/philo-micas/sonic-swss/blob/d58a143c67099c4d73522b01b7bc7f220797e518/cfgmgr/vrrpmgr.cpp#L134-L154) is valid or ever triggered, but I don't see any reason to manage that part from the `vrrpmgr` and to put that code there.
> > > Apart of that, and as also other users suggested, I also believe that warm reboot can be covered by this HLD.
> > > Many thanks and regards.
> > 
> > 
> > Since we need to transmit vip and vmac to OA, appl db need them. and the whole proposal support cli configuration, so a mgrd is needed to do the data transmission from config db to appl db (since other protocol like keepactive will config function direct to kernel, and kernel send data to vrrpsyncd which will transmit data into appl db), And why we not do it in vrrpd, since it will need some system tool to set vip or vmac, like iproute2, but other OS may support ifupdown2, as FRR is open source, will be much more compatible not do it in vrrpd.
> 
> Hi @philo-micas, I understand your point of view and I will certainly take the time to dig deeper - many thanks for your answer and for your input.
> 
> I'm ok with `vrrpsyncd`, I agree that it's necessary.
> 
> Regarding my concern, and maybe due a lack of knowledge, I'm still not sure how `vrrpmgr` will be able to add and delete VIPs into VMAC interfaces in a efficient way without knowing the VRRP's machine state, which is managed by `FRR/vrrpd` through the protocol implementation.
> 
> Maybe, renaming `vrrpmgr` into `macvlanmgr` and sharing it across the whole SONiC architecture should be evaluated and [as also other users pointed in the past](https://github.com/sonic-net/SONiC/pull/475#issuecomment-548054543) it could help to bring more features in the future.
> 
> Regards,

@nser77 Thanks, it is a good idea, I have accepted and updated HLD. @adyeung help forward it, thanks!

vvbrcm reviewed (approved):
These changes look fine. Approved.

vvbrcm reviewed (approved):
Changes look fine. Approved.

philo-micas commented:
@prsunny @lguohan @adyeung Please check if more reviewers are needed for this PR, otherwise, please help merge it. Thanks!

adyeung commented:
This HLD has been reviewed in Routing WG, and SONiC Community call. I will merge the HLD if there is no further comments by the end of the week

nmoray commented:
@adyeung As per the HLD, I am unable to find the changes related to vrrpsyncd and vrrporch in any of your PRs. Can you please point me to respective PR?

philo-micas commented:
> @philo-micas As per the HLD, I am unable to find the changes related to vrrpsyncd and vrrporch in any of your PRs. Can you please point me to the respective PR? cc: @venkatmahalingam

@nmoray The vrrporch and vrrpsyncd sessions have already been included in this HLD. As for the code PR related to them, let's move to the SWSS code PR: https://github.com/sonic-net/sonic-swss/pull/3106. Thanks!

zhangyanzhao commented:
code PRs are not merged, move to backlog
